### PR TITLE
Feature: Flashcard

### PR DESCRIPTION
**Pull request by gromkii**
Closes [PATH-516](https://gohypergiant.atlassian.net/browse/PATH-516)

## ✅ Pull Request Checklist

- [ ] Included link to corresponding [GitHub Issue](https://github.com/gohypergiant/standard-toolkit/issues).
- [x] The commit message follows [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) [extended](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type) guidelines.
- [x] Added/updated unit tests and storybook for this change (for bug fixes / features).
- [x] Added/updated documentation (for bug fixes / features)
- [x] Filled out test instructions.
- [x] Added changeset (for bug fixes / features).

## 📝 Test Instructions

- Run storybook preview locally or check PR build.
- Navigate to new `Flashcard` component tab.

## ❓ Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: .changeset/every-showers-strive.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@accelint/design-toolkit": minor
+---
+
+Includes new Flashcard feature. A display component to quickly summarize information at a glance, with multiple options to show data.
```
Comment (switzerb) [inline .changeset/every-showers-strive.md]:
```suggestion
Adds Flashcard, a display component to quickly summarize information at a glance, with multiple options to show data.
```
nit: this is the the world's most absurd nitpick...but "A display component..." was a sentence fragment and it makes my skin itchy.

File: packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    <FlashcardHero>
+      <FlashcardIdentifier>
+        <FlashcardHeader>Identifier</FlashcardHeader>
+        <FlashcardSubheader>DATA</FlashcardSubheader>
+      </FlashcardIdentifier>
+    </FlashcardHero>
+    <FlashcardSecondaryContainer>
+      <FlashcardSecondaryData>SECONDARY_DATA_01</FlashcardSecondaryData>
+      <FlashcardSecondaryData>SECONDARY_DATA_02</FlashcardSecondaryData>
+      <FlashcardDetailsContainer details={details}/>
```
Comment (switzerb) [inline packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx]: So, we chatted about this a bit already, but my inclination here is to name this `FlashcardDetails` and then have it compose in the `DetailsList` here in the story to demonstrate one example of usage -- rather than trying to pass in the data and have the component itself loop over the data. If `FlashcardDetails` is agnostic and serves as a relatively unopinionated container then I think that adds a lot of flexibility.

The flip side is that if the designers *want* to enforce opinionated control, then my suggestion here is a bad one, so we might want to confirm one way or the other.

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    return null;
+  }
+
+  return (
+    <h2 {...rest} className={secondaryData({ className })}>
+      {children}
+    </h2>
+  );
+}
+FlashcardSecondaryData.display = 'FlashcardSecondaryData';
```
Comment (switzerb) [inline packages/design-toolkit/src/components/flashcard/index.tsx]:
```suggestion
FlashcardSecondaryData.displayName = 'FlashcardSecondaryData';
```

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+  return (
+    <div {...rest} className={subHeader({ className })}>
+      {children}
+    </div>
+  );
+}
+FlashcardSubheader.displayName = 'FlashcardSubheader';
+
+export function FlashcardSecondaryContainer(props: FlashcardComponentProps) {
```
Comment (switzerb) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: I don't think you need the word `Container` on this component. `FlashcardSecondary` is descriptive enough I think.

File: packages/design-toolkit/src/components/flashcard/types.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  ComponentPropsWithRef<'div'>;
+
+export type FlashcardProps = FlashcardComponentProps & {
+  isLoading?: boolean;
+};
+
+// TODO: How strict do we need to be with the value types?
+export type FlashcardDetailProps = FlashcardMetaData & {
+  className?: string;
+};
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/types.ts]: this is defined but never used

File: packages/design-toolkit/src/components/flashcard/flashcard.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import {
+  Flashcard,
+  FlashcardDetailsContainer,
+  FlashcardHeader,
+  FlashcardHero,
+  FlashcardIdentifier,
+  FlashcardSecondary,
+  FlashcardSecondaryData,
+  FlashcardSubheader,
+} from '.';
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/flashcard.stories.tsx]: need a story to show FlashcardSecondaryDetails component. That's missing from examples.

File: packages/design-toolkit/src/components/flashcard/flashcard.test.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+function setup(isLoading = false) {
+  const details = [
+    { label: 'key', value: 'value' },
+    { label: 'key', value: 'value' },
+    { label: 'key', value: 'value' },
+    { label: 'key', value: 'value' },
+    { label: 'key', value: 'value' },
+    { label: 'key', value: 'value' },
+    { label: 'key', value: 'value' },
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/flashcard.test.tsx]: nit: these could be made a little more unique for more realistic scenarios

File: packages/design-toolkit/src/components/flashcard/types.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import type { ComponentPropsWithRef, PropsWithChildren } from 'react';
+
+export type FlashcardComponentProps = PropsWithChildren &
+  ComponentPropsWithRef<'div'>;
+
+export type FlashcardProps = FlashcardComponentProps & {
+  isLoading?: boolean;
+};
+
+export type FlashcardDetailContainerProps = {
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/types.ts]: is there a reason not to include the `ComponentPropsWithRef<'div'>` type extension here too?


```suggestion
export type FlashcardDetailContainerProps = ComponentPropsWithRef<'div'> & {
  details?: FlashcardMetaData[];
  classNames?: { /* ... */ };
};
```

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+  const displayDetails = details.slice(0, 5);
+
+  return (
+    <div
+      {...rest}
+      className={detailsContainer({ className: classNames?.container })}
+    >
+      {/* Limit to the first 5 items in details array.*/}
+      {displayDetails.map((item) => {
+        const detailId = uuid();
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: I know my previous comment led to this, and this is a little funky to create new uuids for these on every render. It'll lead to unnecessary re-renders.

A more stable approach would be to require an id in the item data shape. The id should be created when the `details` data is created, not when it's rendered.

File: packages/design-toolkit/src/components/flashcard/types.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    item?: string;
+    label?: string;
+    value?: string;
+  };
+};
+
+export type FlashcardMetaData = {
+  label: string;
+  value: string | number | null;
+};
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/types.ts]: re: my other comment about ids, add those here:


```suggestion
import type { UniqueId } from '@accelint/core'
...
export type FlashcardMetaData = {
  id: UniqueId;
  label: string;
  value: string | number | null;
};
```

**Review by colehart — APPROVED** [review]

File: packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx (hunk 2 above)
Comment (switzerb) [reply packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx]: Just to document the decision, design wants to ensure that the details remain static (not interactive) and we don't allow too much information into the flashcard, so the restriction here is a choice. We can revisit in the future if we need the capability to be broader.

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+  const { children, className, ...rest } = props;
+  const { isLoading } = useContext(FlashcardContext);
+
+  // While loading, don't display.
+  if (isLoading) {
+    return null;
+  }
+
+  return (
+    <h2 {...rest} className={secondaryData({ className })}>
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: doesn't make sense to have an h2 here when we're formatting the text style. swtich to a `<p>` or `<div>` tag

File: packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+# Flashcard
+A quick visual identifier for multiple points of data for a given map entity.
+
+## Example Usage
+
+### Basic
+
+```tsx
+  const details = [
+    { label: 'key', value: 'value', id: unique-id-1 },
```
Comment (ArrayKnight) [inline packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx]: NIT: You should wrap those unique ids in quotes

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+/**
```
Comment (ArrayKnight) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: Please move this comment below the imports

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+  return (
+    <div {...rest} className={identifier({ className })}>
+      {children}
+    </div>
+  );
+}
+FlashcardIdentifier.displayName = 'FlashcardIdentifier';
+
+export function FlashcardHeader(props: FlashcardComponentProps) {
```
Comment (ArrayKnight) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: You can simplify this component to just be `export const FlashcardHeader = Heading`, then the subheader can be `export const FlashcardSubheader = Text`, and then implement the context provider in the hero to pass the classes down. This helps to eliminate wrapper components that do almost nothing

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+  return (
+    <div {...rest} className={subHeader({ className })}>
+      {children}
+    </div>
+  );
+}
+FlashcardSubheader.displayName = 'FlashcardSubheader';
+
+export function FlashcardSecondary(props: FlashcardComponentProps) {
```
Comment (ArrayKnight) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: I think you can get rid of this component entirely. Since it seem to be solely responsible for a bit of layout, that can be handled by the Flashcard itself. I would potentially reach for a 5 column grid. The 1st, 3rd, and 5th columns being gutters, 2nd and 4th being named. Hero spans all columns, secondary data spans 2-4, and the details list is on the two named columns label and value

Comment (ArrayKnight) [reply packages/design-toolkit/src/components/flashcard/index.tsx]: Scratch that, that's too complicated. Just apply the padding to the Flashcard and use negative margin on the hero and the details list

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+  return (
+    <div {...rest} className={flashcardData({ className })}>
+      {children}
+    </div>
+  );
+}
+FlashcardAdditionalData.displayName = 'FlashcardAdditionalData';
+
+export const FlashcardDetailsList = (props: FlashcardDetailsListProps) => {
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: missing displayName for this one. Also, every other component uses `function` sytax versus `const`/arrow. Worth consistency of one pattern, so I'd suggest updating this to be `function`

File: packages/design-toolkit/src/components/flashcard/types.ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+export type FlashcardComponentProps = PropsWithChildren &
+  ComponentPropsWithRef<'div'>;
+
+export type FlashcardProps = FlashcardComponentProps & {
+  isLoading?: boolean;
+};
+
+export type FlashcardDetailsListProps = FlashcardComponentProps &
+  Partial<DetailsListProps>;
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/types.ts]: ?double checking my understanding: are there specific things you want from DetailsListProps or any overlapping types between these two types?

File: packages/design-toolkit/src/components/flashcard/flashcard.test.tsx [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+    expect(screen.getByText('SECONDARY_DATA_02')).toBeDefined();
+  });
+
+  it('should only show 5 additional details', () => {
+    setup();
+
+    // First item to be hidden by selector nth-of-type
+    const detailEntries = document.getElementsByClassName(
+      'nth-of-type-[n+6]:hidden',
+    );
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/flashcard.test.tsx]: I think there's a much easier way to test this - you could instead look at what is actually rendered on the screen and test that the labels of the items past 5 .not.toBeInTheDocument()

generally speaking, you shouldn't be using document queries in your tests. Instead you should stick to the RTL queries off of screen.

For this one, since you're asserting something is not in the document, it should be `expect(screen.queryByText('STATE')).not.toBeInTheDocument()`

File: packages/design-toolkit/src/components/flashcard/flashcard.test.tsx [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+    // Should not render header text.
+    const header = screen.queryByText('IDENTIFIER');
+    expect(header).toBeNull();
+
+    const hero = screen.getByTestId('hero');
+    expect(hero).toBeDefined();
+
+    // Two skellington components
+    expect(hero.childElementCount).toEqual(2);
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/flashcard.test.tsx]: there should be some kind of grabbable prop on the skeleton components I think to assert they're specifically what the children are:
`screen.getByTestId()` is what I would grab, and I would add a data-testid attribute to the skeleton components if they dont' have any identifiable other attributes in the screen.debug() you could grab.

File: packages/design-toolkit/src/components/flashcard/flashcard.test.tsx [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    expect(secondaryContainer.childElementCount).toBe(5);
+  });
+
+  it('should not show secondary data field while loading', () => {
+    setup(true);
+    const header = screen.queryByText('IDENTIFIER');
+    expect(header).toBeNull();
+    const secondaryData = screen.queryByText('SECONDARY_DATA');
+    expect(secondaryData).toBeNull();
+  });
```
[the remark this answers is not in the document]
Comment (colehart) [reply packages/design-toolkit/src/components/flashcard/flashcard.test.tsx]: more tests could fill this out:
Test for FlashcardHeader rendering independently
Test for FlashcardSubheader rendering independently
Test that FlashcardDetailLabel/Value show skeletons when loading
Edge case: What happens if Flashcard children are used outside Flashcard context?

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+      <div {...rest} className={container({ className })}>
+        {children}
+      </div>
+    </FlashcardContext.Provider>
+  );
+}
+Flashcard.displayName = 'Flashcard';
+
+export function FlashcardHero(props: FlashcardComponentProps) {
+  const { children, className, ...rest } = props;
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: you're not actually using the className below. Should you be?

File: packages/design-toolkit/src/components/flashcard/types.ts (hunk 17 above)
Comment (gromkii) [reply packages/design-toolkit/src/components/flashcard/types.ts]: This inclusion was due to a type conflict when using `{...rest}` in the `DetailsList` component. It allows us to accept any `DetailsList` props if needed, but they won't be required.

File: packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+      <FlashcardHeader>Identifier</FlashcardHeader>
+      <FlashcardSubheader>DATA</FlashcardSubheader>
+    </FlashcardHero>
+    <FlashcardAdditionalData>SECONDARY_DATA_01</FlashcardAdditionalData>
+    <FlashcardAdditionalData>SECONDARY_DATA_02</FlashcardAdditionalData>
+    <FlashcardDetailsList>
+      {details.map((detail) => (
+        <Fragment key={detail.id}>
+          <FlashcardDetailsLabel>{detail.label}</FlashcardDetailsLabel>
+          <FlashcardDetailsValue>{detail.value}</FlashcardDetailsValue>
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/flashcard.docs.mdx]: As exported, these are typos
```suggestion
          <FlashcardDetailLabel>{detail.label}</FlashcardDetailLabel>
          <FlashcardDetailValue>{detail.value}</FlashcardDetailValue>
```

Worth considering if you want DetailsList to be DetailList to keep it all singular or if you want to pluralize the Label and Value names

File: packages/design-toolkit/src/components/flashcard/index.tsx [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+  const { isLoading } = useContext(FlashcardContext);
+
+  if (isLoading) {
+    return (
+      <div {...rest} className={clsx(styles.hero, className, 'gap-s')}>
+        <Skeleton
+          className={clsx(styles.skeleton)}
```
Comment (colehart) [inline packages/design-toolkit/src/components/flashcard/index.tsx]: no need for clsx when only one class present

```suggestion
          className={styles.skeleton}
```

**Review by colehart — DISMISSED** [review]
Looks good!

**Comment by joshuamora-programmed** [comment]
looks good

**Review by joshuamora-programmed — APPROVED** [review]
